hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -3203,10 +3206,90 @@ func reconcileMachineApproverDeployment(deployment *appsv1.Deployment, hc *hyper
 	return nil
 }
 
-const oidcDocumentsFinalizer = "hypershift.io/aws-oidc-discovery"
+const (
+	oidcDocumentsFinalizer         = "hypershift.io/aws-oidc-discovery"
+	serviceAccountSigningKeySecret = "sa-signing-key"
+	serviceSignerPublicKey         = "service-account.pub"
+	jwksURI                        = "/openid/v1/jwks"
+	discoveryTemplate              = `{
+	"issuer": "%s",
+	"jwks_uri": "%s%s",
+	"response_types_supported": [
+		"id_token"
+	],
+	"subject_types_supported": [
+		"public"
+	],
+	"id_token_signing_alg_values_supported": [
+		"RS256"
+	]
+}`
+)
+
+type oidcGeneratorParams struct {
+	issuerURL string
+	pubKey    []byte
+}
+
+type oidcDocumentGeneratorFunc func(params oidcGeneratorParams) (io.ReadSeeker, error)
+
+func generateConfigurationDocument(params oidcGeneratorParams) (io.ReadSeeker, error) {
+	return strings.NewReader(fmt.Sprintf(discoveryTemplate, params.issuerURL, params.issuerURL, jwksURI)), nil
+}
+
+type KeyResponse struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+func generateJWKSDocument(params oidcGeneratorParams) (io.ReadSeeker, error) {
+	block, _ := pem.Decode(params.pubKey)
+	if block == nil || block.Type != "RSA PUBLIC KEY" {
+		return nil, fmt.Errorf("failed to decode PEM block containing RSA public key")
+	}
+	pubKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public key")
+	}
+	rsaPubKey, ok := pubKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("public key is not RSA")
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(block.Bytes)
+	hash := hasher.Sum(nil)
+	kid := base64.RawURLEncoding.EncodeToString(hash)
+
+	var keys []jose.JSONWebKey
+	keys = append(keys, jose.JSONWebKey{
+		Key:       rsaPubKey,
+		KeyID:     kid,
+		Algorithm: string(jose.RS256),
+		Use:       "sig",
+	})
+
+	jwks, err := json.MarshalIndent(KeyResponse{Keys: keys}, "", "  ")
+	if err != nil {
+		return nil, err
+	}
 
-func oidcPublicDocumentPaths() []string {
-	return []string{"/.well-known/openid-configuration", "/openid/v1/jwks"}
+	return bytes.NewReader(jwks), nil
+}
+
+func oidcDocumentGenerators() map[string]oidcDocumentGeneratorFunc {
+	return map[string]oidcDocumentGeneratorFunc{
+		"/.well-known/openid-configuration": generateConfigurationDocument,
+		jwksURI:                             generateJWKSDocument,
+	}
+}
+
+func hasKeys(secret *corev1.Secret, keys ...string) bool {
```

Comment:
Can this function be deleted and replaced by:

```go
sets.StringKeySet(secret.Data).HasAll("a", "b", "c", ...)
```

Reply:
It can be.  Just copy-pasted from here https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/pki/util.go#L71-L78

Reply:
Then you could delete them both... think of the LOC deletion points you'd get!